File: StreamProcessorService/src/main/java/com/stream/processor/validation/ValidationResult.java
package com.stream.processor.validation;


import lombok.Value;

import java.util.ArrayList;
import java.util.List;

@Value
public class ValidationResult {
    boolean valid ;
    List<String> errors ;

    public static ValidationResult valid(){
        return new ValidationResult(true , new ArrayList<>()) ;
    }

    public static ValidationResult invalid(String error){
        List<String> errors = new ArrayList<>();
        errors.add(error);
        return new ValidationResult(false, errors);
    }

    public ValidationResult combine(ValidationResult other){
        if(this.valid && other.valid){
            return valid() ;
        }
        List<String> allErrors = new ArrayList<>(this.errors);
        allErrors.addAll(other.errors);
        return new ValidationResult(false, allErrors);
    }
}

File: validation_service/validation_result.py
"""Validation result with errors"""
from typing import List


class ValidationResult:
    """Validation result with errors"""
    def __init__(self, valid: bool, errors: List[str] = None):
        self.valid = valid
        self.errors = errors or []
    
    @staticmethod
    def valid_result():
        return ValidationResult(True, [])
    
    @staticmethod
    def invalid_result(error: str):
        return ValidationResult(False, [error])
    
    def combine(self, other: 'ValidationResult') -> 'ValidationResult':
        if self.valid and other.valid:
            return ValidationResult.valid_result()
        all_errors = self.errors + other.errors
        return ValidationResult(False, all_errors)

